### ISSUE TITLE
How is num_epochs calculated? 

### ISSUE THREAD

**Issue by specblades** [kohya-ss/sd-scripts#366]
Cant understand, how is num_epochs calculated. 
I use, for example, 16 repeats, 5 epochs (in gui), 16 grad accum steps. And i get 40 epoch.
Why batches per epoch is 32?
What? 😢
Help me out pls
<img width="286" alt="image" src="https://user-images.githubusercontent.com/113993746/229115467-81d421d9-5251-4a65-8898-0b6ebd9c577b.png">

**Comment by kgonia** [comment]
what value for **max_train_steps** did you use?

**Comment by specblades** [comment]
> The number of batches per epoch is `32` in this case. Because the number of images to train is `32` (`16` train images and `16` reg images.)
> 
> However, the number of actual update steps per epoch is `2`. Because the gradient accumulation steps is `16`, the update (back propagation) per epoch is `32/16=2`.
> 
> Therefore, `total optimization steps` will be `10`, because it is calculated by `max_train_epochs * update_steps`...

But I uploaded 28 reg images. Does that mean the maximum reg image equals the maximum train image?

**Comment by specblades** [comment]
> Yes, the reg images are repeated until the number of the training images, to balance the numbers.

Will only the "N=train images" images be utilized from the entire reg folder? Or will all the images be used sequentially?
